### PR TITLE
Update to babel 6 #78

### PR DESCRIPTION
**Pull request by lecstor**
- tests pass with React 13 & 14
- npm run travis produces coverage
- npm link works

### CODE REVIEW

**Comment by lecstor** [comment]
hrmm bugga. iojs-v3.3 `TypeError: Object.assign is not a function`
I will have to come back to this later.

**Comment by ljharb** [comment]
@lecstor `Object.assign` isn't available natively in all node versions.

File: package.json [hunk 1]
```diff
@@ -51,8 +51,15 @@
...
-    "babel": "^5.8.21",
+    "babel-cli": "^6.3.17",
+    "babel-core": "^6.3.21",
     "babel-eslint": "^4.1.4",
+    "babel-istanbul": "^0.5.9",
+    "babel-preset-es2015": "^6.3.13",
+    "babel-preset-react": "^6.3.13",
+    "babel-preset-stage-0": "^6.3.13",
+    "babel-register": "^6.3.13",
+    "babel-runtime": "^6.3.19",
```
Comment (ljharb) [inline package.json]: It seems very unfortunate to require `babel-runtime` in a reusable module. Why is this necessary?

File: src/index.js [hunk 2]
```diff
@@ -3,7 +3,8 @@ import Sinon from 'sinon';
 import ReactWrapper from './ReactWrapper';
 import ShallowWrapper from './ShallowWrapper';
 import { onPrototype } from './Utils';
-import { renderToStaticMarkup } from './react-compat';
+import reactCompat from './react-compat';
+const {renderToStaticMarkup} = reactCompat;
```
Comment (ljharb) [inline src/index.js]: please preserve the spaces inside the object destructuring brackets

Comment (lelandrichardson) [inline src/index.js]: curious, but why were these changed from imports to imports + destructure?  Is there value to doing so? is this just a style preference?

File: .babelrc [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+  presets: ["react", "es2015", "stage-0"]
```
Comment (goatslacker) [inline .babelrc]: Use this preset instead https://github.com/airbnb/babel-preset-airbnb

File: package.json (hunk 1 above)
Comment (lecstor) [inline package.json]: thanks, you are correct, it is not required and will be removed.

File: src/index.js (hunk 2 above)
Comment (lecstor) [inline src/index.js]: how easily we come to rely on linting in the editor.. I will update eslint-config-airbnb and fix linting.

Babel6 has changed the way bare objects are exported for compatability reasons meaning the properties of the exported object are actually in a `default` property.

```
TypeError: (0 , _reactCompat.renderToStaticMarkup) is not a function
      at render (index.js:8:30)
      at matchesRender (ReactWrapper-spec.js:6:9)
      at Context.<anonymous> (ReactWrapper-spec.js:540:7)
```

The alternative would be to use require:
`const { renderToStaticMarkup } = require('./react-compat').default;`

I'm not aware of any other ES6 options.

**Comment by lecstor** [comment]
`babel-plugin-transform-object-assign` fixes the iojs issues. I guess Babel5 was doing this by default.
What is the preference here?
- add the plugin to enzyme
- add the plugin to the airbnb preset
- replace the single use of `Object.assign` in ReactWrapperComponent.jsx with an iojs compatible equivalent

**Comment by ljharb** [comment]
@lecstor the single use could be replaced with `object.assign` usage which would be compatible with even node 0.4 if necessary - that seems simplest rather than adding another transform.

File: package.json [hunk 4]
```diff
@@ -47,20 +47,25 @@
   "license": "MIT",
   "dependencies": {
     "cheerio": "^0.19.0",
+    "object-assign": "^4.0.1",
```
Comment (ljharb) [inline package.json]: please use `object.assign` instead of `object-assign` - it's more spec-compliant.

File: src/index.js (hunk 2 above)
Comment (ljharb) [inline src/index.js]: ES6 imports indeed can't be destructured inline, per spec - babel 5 violates the spec in this regard. babel 6 violates it in a different way.

Since the ES6 loader spec doesn't yet exist, ES6 modules don't yet exist, so technically, nobody should be using import/export anywhere under any circumstances. Since a number of people have chosen to use the syntax anyways, a number of compromises have to be made depending on the situation. This is appropriate here.

**Comment by lecstor** [comment]
sorry about that. I haven't needed to use it before (obviously 8).
Is the `shim` method ok? there appear to be a few ways to use it..

**Comment by ljharb** [comment]
Since this isn't a top-level app, only a reusable module, modifying the global environment in any way is unacceptable - so please don't use the `shim` method, just use the module as a pure function.

**Comment by lecstor** [comment]
done. I think. I'm not entirely sure if that is the correct usage. Reading the source it seems it should be fine, but the readme doesn't mention that usage style at all. The examples all call shim or getPolyfill, and assign the result to shimmedAssign mostly, but use `assign` as the function (from the first example) so I found them a little confusing.

**Comment by ljharb** [comment]
@lecstor that looks great, thanks! The readme on https://www.npmjs.com/package/object.assign has the examples under "most common usage".

**Comment by lecstor** [comment]
excellent, thank you @ljharb

File: src/MountedTraversal.js [hunk 5]
```diff
@@ -6,13 +6,16 @@ import {
...
-import {
+import reactCompat from './react-compat';
+const {
   isDOMComponent,
   isCompositeComponent,
   isCompositeComponentWithType,
   isElement,
   findDOMNode,
-} from './react-compat';
+} = reactCompat;
```
Comment (lelandrichardson) [inline src/MountedTraversal.js]: was this changed for a reason?

File: src/__tests__/ReactWrapper-spec.js [hunk 6]
```diff
@@ -506,18 +506,20 @@ describeWithDOM('mount', () => {
...
       }
       matchesRender(<Foo items={['abc', 'def', 'hij']} />);
       matchesRender(
-        <Foo items={[
-          <i key={1}>abc</i>,
-          <i key={2}>def</i>,
-          <i key={3}>hij</i>,
-        ]} />
+        <Foo items=
+          {[
```
Comment (lelandrichardson) [inline src/__tests__/ReactWrapper-spec.js]: this indentation seems strange. was this required by the linter or something?

File: src/MountedTraversal.js (hunk 5 above)
Comment (ljharb) [inline src/MountedTraversal.js]: babel 5 incorrectly allows destructuring on a default import - the spec does not.

Comment (lelandrichardson) [inline src/MountedTraversal.js]: this makes me sad :/

Comment (goatslacker) [inline src/MountedTraversal.js]: Why don't we fix this in `react-compat.js`?

Just `export` all these individually.

Comment (lelandrichardson) [inline src/MountedTraversal.js]: :+1: i agree that would be better

File: src/__tests__/ReactWrapper-spec.js (hunk 6 above)
Comment (lecstor) [inline src/__tests__/ReactWrapper-spec.js]: it does, doesn't it.. but yes, it seemed the nicest way I could make the linter happy.

File: src/MountedTraversal.js (hunk 5 above)
Comment (lecstor) [inline src/MountedTraversal.js]: PR updated accordingly.

File: src/__tests__/ReactWrapper-spec.js (hunk 6 above)
Comment (ljharb) [inline src/__tests__/ReactWrapper-spec.js]: you want this:
```
<Foo
  items={[
    …
  ]}
/>
```

Comment (lecstor) [inline src/__tests__/ReactWrapper-spec.js]: ok, ta, will update when I get a chance.

**Comment by lelandrichardson** [comment]
Hey @lecstor, we are wanting to merge this now....

Can you rebase and squash your commits?  Thanks!

**Comment by lecstor** [comment]
Hi @lelandrichardson, I had to make a final change to update the test commands to use `babel-core/register`. Should be suitably rebased and squashed now.
